helix-view/src/register.rs

```diff
@@ -0,0 +1,235 @@
+use std::{borrow::Cow, collections::HashMap, iter};
+
+use anyhow::Result;
+use helix_core::NATIVE_LINE_ENDING;
+
+use crate::{
+    clipboard::{get_clipboard_provider, ClipboardProvider, ClipboardType},
+    document::SCRATCH_BUFFER_NAME,
+    Editor,
+};
+
+/// A key-value store for saving values.
+///
+/// Each register corresponds to a `char`. Most chars can be used to store any set of
+/// values but a few chars are "special registers". Special registers have unique
+/// behaviors when read or written to:
+///
+/// * Black hole (`_`): all values read and written are discarded
+/// * Selection indices (`#`): index number of each selection starting at 1
+/// * Selection contents (`.`)
+/// * Document path (`%`): filename of the current buffer
+/// * System clipboard (`*`)
+/// * Primary clipboard (`+`)
+#[derive(Debug)]
+pub struct Registers {
+    inner: HashMap<char, Vec<String>>,
+    clipboard_provider: Box<dyn ClipboardProvider>,
+}
+
+impl Default for Registers {
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+            clipboard_provider: get_clipboard_provider(),
+        }
+    }
+}
+
+// Some special registers must allocate their values while others and regular
+// registers can hand out borrowed values.
+type RegisterValues<'a> = Box<dyn ExactSizeIterator<Item = Cow<'a, str>> + 'a>;
+
+impl Registers {
+    pub fn read<'a>(&'a self, name: char, editor: &'a Editor) -> Option<RegisterValues<'a>> {
+        match name {
+            '_' => Some(Box::new(iter::empty())),
+            '#' => {
+                let (view, doc) = current_ref!(editor);
+                let selections = doc.selection(view.id).len();
+                // ExactSizeIterator is implemented for Range<usize> but
+                // not RangeInclusive<usize>.
+                Some(Box::new(
+                    (0..selections).map(|i| i.saturating_add(1).to_string().into()),
```

Comment:
```suggestion
                    (1..=selections).map(|i| i.to_string().into()),
```

nit this reads slightly better and there is no need to saturate. Since we can never get a value larger than selections (which it already a valid usize

Reply:
I would prefer this but ExactSizeIterator isn't implemented for `RangeInclusive<usize>` so we use a regular `Range<usize>` as a workaround. I assume it isn't implemented because the `ExactSizeIterator::len` could hypothetically be `usize::MAX + 1`